source/core/slang-dictionary.h

```diff
@@ -135,15 +135,17 @@ namespace Slang
 			}
 
 		};
-		inline int GetHashPos(TKey& key) const
+        template<typename KeyType>
+		inline int GetHashPos(KeyType& key) const
```

Comment:
We have a few operations on `Dictionary` that allow these kinds of operations, where the user can pass something that isn't actually of the correct key type. This choice seems to be designed around things like smart pointers where the value you use as the key might not be exactly what the dictionary stores.

The concern I have is that this approach seems to allow a value to be used for the key that is different from the actual key type, but is convertible to it, and the implicit constraint is that the hash of the value before/after conversion should not be changed.

What would happen if I had a `Dictinary<int, int>` and then I tried to use a `float` as a key? Hashing my key as a `float` instead of an `int` would seemingly lead to different values. How does the code here protected against that kind of stuff?